[Lim Peiyuan] iP

Py0000 opened the pull request:
### DukePro

> Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
<!-- blank line -->
----
<!-- blank line -->
If you Java programmer, you can use it to practice Java too. Here's the  `main` method:

```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level-1, Implement sketal version of Duke.

Commit:
Level 2, Add the ability to store text and display text.

Commit:
Level 3, Add Task class and ability to mark tasks as done.

Commit:
Updated Method to print task as string.

Commit:
Level 3, Updated with the new changes in Task Class.

Commit:
Level 3, Task Class updated.

Commit:
Level 4

Commit:
Level 4

Commit:
Level 4

Commit:
Level 4, Add ToDos, Events, Deadlines and Edit codes.

Commit:
Level 4, A-TextUiTesting Done, Fix Bugs

Commit:
Level 4, A-TextUiTesting Done

Commit:
Level 4, A-TextUiTesting Done

Commit:
Level 5, Add Exception Handling

Commit:
Level 5. Add New Exceptions

Commit:
Level 6, Add delete task

Commit:
Level 6, Done A-TextUiTesting

Commit:
gitignore

Commit:
Level 6, Minor edits

Commit:
Level 5, Add JavaDoc/Comments

Commit:
Add Level 7 save functionally.

Commit:
Revert "Add Level 7 save functionally."

This reverts commit 9aefefb5b4af857e230751691ad21baa755def33.

Commit:
Add Level 7 save feature.

Commit:
Uppdate Duke Class with Level 7 feature

Commit:
Extract into exceptions package

Commit:
Extract into task package

Commit:
Extract into packages

Commit:
Add data file for the chat bot.

Commit:
Update Duke with Packages.

Commit:
Add Dates and Times.

Commit:
Merge branch 'branch-Level-7'

Commit:
Merge branch 'branch-Level-8'

Commit:
Fix bugs after merging the branches

Commit:
Add Ui class to handle interactions of chat bot with user.

Commit:
Extract methods that deals with interaction with user to Ui class.

Commit:
Update code for getDueDate method.

Commit:
Fix bugs after merging branch-Level-7 and branch-Level-8

Commit:
Add TaskList class to deal with manipulation of tasks.

Commit:
Add Parser Class to deal with interpreting user input.

Commit:
Add TaskList class to handle manipulation of task.

Commit:
Extract into Packages

Commit:
Update code to catch DateTimeParseException.

Commit:
Update code to catch DateTimeParseException.

Commit:
Add JUnit Test

Commit:
Add JavaDocs.

Commit:
Update code according to Coding Standard

Commit:
Update code according to Coding Standard

Commit:
Add find functionality

Commit:
Add data file for chat bot

Commit:
Update code according to coding standard

Add "this" keyword to reference class fields.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Resolve Merge Conflict

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Ui.java

Commit:
Fix code

Change method names in Ui class and DoneCommand class.

Commit:
Fix JavaDoc

Commit:
Complete A-TextUiTesting after Level 9 feature

Commit:
Remove irrelevant file

Commit:
Merge branch 'add-gradle-support'

Commit:
Update gradle files

Commit:
Add checkstyle.xml file

Commit:
Update code after running checkstyle

Commit:
Fix checkstyle errors

Commit:
Add image resources for GUI

User Icon from: https://surveyinggroup.com/power-line-corridor-mapping-with-lidar/batman-icon-2/
Bot Icon from: https://imgbin.com/png/fYu4Dyqt/the-iron-man-war-machine-png

Commit:
Update gradle for JavaFx

Commit:
Add images for GUI

User image from: https://lh3.googleusercontent.com/proxy/3eOg-6TPXMBz7GszN2hWV-k2Oez5sujJJJ01CUqGEYIW57uok3_nyEK5TmCl8p1lcVzK4C0mz5szfwXRcF4PdRyw5NjoisRT--70qSuBU1XUgjnB01bqWDDmyCc
Bot image from: https://images-wixmp-ed30a86b8c4ca887773594c2.wixmp.com/f/ec0429cf-1471-466c-b958-f3b1faa7896b/d5z37bp-e5528f07-d513-4b67-87f1-8385e3fd4c91.png?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ1cm46YXBwOjdlMGQxODg5ODIyNjQzNzNhNWYwZDQxNWVhMGQyNmUwIiwiaXNzIjoidXJuOmFwcDo3ZTBkMTg4OTgyMjY0MzczYTVmMGQ0MTVlYTBkMjZlMCIsIm9iaiI6W1t7InBhdGgiOiJcL2ZcL2VjMDQyOWNmLTE0NzEtNDY2Yy1iOTU4LWYzYjFmYWE3ODk2YlwvZDV6MzdicC1lNTUyOGYwNy1kNTEzLTRiNjctODdmMS04Mzg1ZTNmZDRjOTEucG5nIn1dXSwiYXVkIjpbInVybjpzZXJ2aWNlOmZpbGUuZG93bmxvYWQiXX0.Xr3PcfDk1CShJFA5KA_defLEWCE2iWeC8jysznsRkYc

Commit:
Add GUI implementation into code

Commit:
Merge remote-tracking branch 'origin/branch-Level-10' into branch-Level-10

Commit:
Update GUI

Add function to close the application 1.25secs after user types 'bye' command

wowsiddanth commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public String getResponse(String input) {
+        try {
+            String exitTag = "bye";
+            if (input.equalsIgnoreCase(exitTag)) {
+                new Thread() {
+                    public void run() {
+                        try {
+                            Thread.sleep(1250);
+                        } catch (Exception e) {
+                            e.printStackTrace();
+                        } finally {
+                            System.exit(0);
+                        }
+                    }
+                }.start();
+
```
There is some deep nesting here, maybe you can try simplifying it to be less "arrow-head" like.

IceWizard4902 replied:
Not really a coding violation, but I do agree with Siddanth. Improving readability here would be nice.

Py0000 replied:
Thank you. Will change this part in the next commit.

wowsiddanth commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.io.IOException;
+
+import duke.gui.MainWindow;
+import javafx.application.Application;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.Scene;
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
```
Minor nit-pick: I think the duke import should be separated from the JavaFX ones. Please do check if I am correct as I am not sure as well.

Py0000 replied:
I think you are right. Thank you for pointing it out!

wowsiddanth commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class Event extends Task {
+    public static final String TAG = "E";
+    private LocalDate at;
```
Maybe the name can be more descriptive.

IceWizard4902 replied:
Again, not coding violation but I do agree with Siddanth. The name can be more descriptive or you can add comments to better describe this

Py0000 replied:
Thank you. Will improve on it in the next commit.

wowsiddanth commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,205 @@
+        }
+
+        try {
+            return Integer.parseInt(details[1]) - 1;
```
Maybe abstract this out.

Py0000 replied:
Thank you for the suggestion. Will try to abstract it out.

wowsiddanth reviewed (approved):
I loved your program! It was logical and consistent.

Other than a few nitpicks, it LGTM 😀

IceWizard4902 reviewed (commented):
Overall, it looks great and nice use of OOP!

IceWizard4902 commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,52 @@
+public class TaskList {
+    private ArrayList<Task> items;
+
+    public TaskList() {
```
Maybe you can add in Javadocs here for the public methods

Py0000 replied:
Thank you for pointing it out. Will add in the javadocs here!

IceWizard4902 reviewed (approved):
Everything looks great! I actually learn how you organize the code.

As Siddanth pointed out, there are a few things to nitpick. But overall, nice job!

Commit:
Remove duplicated UI.java file.

Update code after removal of UI.java file.

Commit:
Add suppression.xml file

Commit:
Update code with the suggestions given during PR review

Commit:
Update code

Additional line breaks in between 2 lines of codes within the same method

Remove the extra line break in between the 2 lines of codes

Removing the extra line break will help to standardise the code and improve readability

Commit:
Update code for a new situation

The application will print out an empty message if the task list is
empty.

Add a new if-block to check if the tasklist to be displayed is empty,
if it is empty the application will print out an empty list message
instead of an empty list. Empty list message will inform the user that
there is no task in his/her list at the moment.

Doing so will help to improve the user experience of the application, as
it notifies the user that he/she does not have any task currently.
Instead of displaying an empty list. which might confuse to user, as it
may seem as an error from the application.

Commit:
Update code

Commit:
Use assertions

To document important assumptions about files to be loaded for the application.

These are the conditions that have to be true for the
application to work properly.

Check if the files to be loaded exists and are loaded correctly,

Assertions are used because if the files fails to load or fails to
exist, it is considered to be the programmer mistake.

Py0000 committed:
Merge pull request #1 from Py0000/branch-A-Assertions

Use assertions

Commit:
Fix error in assert statement

Py0000 committed:
Merge pull request #2 from Py0000/branch-A-Assertions

Fix error in assert statement

Commit:
Fix logic in code

Commit:
Revert "Fix logic in code"

This reverts commit 3d7764b99618a5e78b5bfa61b497f7b6eaaef428.

Commit:
Fix logic in code

Py0000 committed:
Merge pull request #3 from Py0000/branch-A-Assertions

Fix logic in code

Commit:
Remove unnecessary files

Commit:
Add Junit Tests

Commit:
Improve Code Quality

Code Quality not up to standard.

Improving code quality helps to improve readability and understandability.

Followed CS2103T Code Quality Standard and Convention.

Py0000 committed:
Merge pull request #4 from Py0000/branch-A-CodeQuality

Improve Code Quality

Commit:
Add update extension

The user cannot modify or edit the existing task in the list directly.
The only way a user can 'edit' an existing task is for him to delete
and re-input the updated task.

This extension allows the user to be able to directly
modify the existing task by making use of 'update'
command.

This is to make the application more user-friendly as
it now supports another useful feature.

Commit:
Minor touch up to code quality.

Commit:
Update Javadoc for UpdateCommand Class

Commit:
Add CI

Commit:
Update gradle file

Commit:
Improve colors of GUI slightly

Commit:
Replace images used.

UserIcon:
Image by André Santana AndreMS from Pixabay
https://pixabay.com/vectors/captain-america-marvel-hero-6192855/

BotIcon:
Image by Ega Maulana from Pixabay
https://pixabay.com/vectors/iron-man-chibi-cartoon-superhero-5783522/

Commit:
Update Comments

Commit:
Add a representative screenshot of product

Py0000 committed:
Update README.md

Commit:
Minor UI updates

Commit:
Merge branch 'master' of https://github.com/Py0000/ip

Commit:
Fix issues highlighted by cs2103-bot.

Commit:
Update README.md

Commit:
Improve GUI

Commit:
Minor update to Ui.png
